File: ChessLab/Variants/DuckChessSAN.swift
import ChessKit

/// Notation d'un coup de Duck Chess.
///
/// Proche du SAN classique, à une différence près qui dit la variante : il n'y
/// a ni « + » ni « # », puisqu'il n'y a ni échec ni mat. La capture du roi,
/// elle, se marque d'un « ++ » — une convention lisible pour le coup qui
/// termine la partie.
enum DuckChessSAN {

    static func build(
        move: DuckChessRules.Move, position: Position,
        legalMoves: [DuckChessRules.Move], capturesKing: Bool
    ) -> String {
        guard let piece = position.piece(at: move.from) else { return move.uci }
        let suffix = capturesKing ? "++" : ""

        if piece.kind == .king, abs(move.to.file.number - move.from.file.number) == 2 {
            return (move.to.file.number > move.from.file.number ? "O-O" : "O-O-O") + suffix
        }

        let targetOccupied = position.piece(at: move.to) != nil
        let isEnPassant = piece.kind == .pawn && move.from.file != move.to.file && !targetOccupied
        let isCapture = targetOccupied || isEnPassant

        var body = ""
        if piece.kind == .pawn {
            if isCapture { body += String(move.from.notation.prefix(1)) + "x" }
            body += move.to.notation
            if let promotion = move.promotion { body += "=" + promotion.rawValue }
        } else {
            body += piece.kind.rawValue
            body += disambiguation(piece: piece, move: move, legalMoves: legalMoves, position: position)
            if isCapture { body += "x" }
            body += move.to.notation
        }
        return body + suffix
    }

    private static func disambiguation(
        piece: Piece, move: DuckChessRules.Move,
        legalMoves: [DuckChessRules.Move], position: Position
    ) -> String {
        let others = legalMoves.compactMap { candidate -> Square? in
            guard candidate.from != move.from, candidate.to == move.to,
                  let other = position.piece(at: candidate.from),
                  other.kind == piece.kind, other.color == piece.color
            else { return nil }
            return candidate.from
        }
        guard !others.isEmpty else { return "" }
        if !others.contains(where: { $0.file == move.from.file }) {
            return String(move.from.notation.prefix(1))
        }
        if !others.contains(where: { $0.rank == move.from.rank }) {
            return String(move.from.notation.suffix(1))
        }
        return move.from.notation
    }
}

File: android/app/src/main/java/com/chesslab/variants/DuckChessSan.kt
package com.chesslab.variants

import chesskit.Piece
import chesskit.Position
import chesskit.Square
import kotlin.math.abs

/**
 * La notation d'un coup de Duck Chess. Pendant de `DuckChessSAN.swift`.
 *
 * Proche du SAN ordinaire, à une différence près, et c'est elle qui dit la
 * variante : ni « + » ni « # », puisqu'il n'y a ni échec ni mat — le canard
 * peut boucher la seule case qui sauvait le roi, et la partie se gagne en
 * PRENANT le roi. Cette prise-là se marque « ++ », convention lisible pour le
 * coup qui termine la partie.
 */
object DuckChessSan {

    fun build(
        move: DuckChessRules.Move,
        position: Position,
        legalMoves: List<DuckChessRules.Move>,
        capturesKing: Boolean,
    ): String {
        val piece = position.piece(move.from) ?: return move.uci
        val suffix = if (capturesKing) "++" else ""

        if (piece.kind == Piece.Kind.king && abs(move.to.file.number - move.from.file.number) == 2) {
            return (if (move.to.file.number > move.from.file.number) "O-O" else "O-O-O") + suffix
        }

        val occupied = position.piece(move.to) != null
        val enPassant = piece.kind == Piece.Kind.pawn && move.from.file != move.to.file && !occupied
        val capture = occupied || enPassant

        val body = StringBuilder()
        if (piece.kind == Piece.Kind.pawn) {
            if (capture) body.append(move.from.file.letter).append("x")
            body.append(move.to.notation)
            move.promotion?.let { body.append("=").append(it.notation) }
        } else {
            body.append(piece.kind.notation)
            body.append(disambiguation(piece, move, legalMoves, position))
            if (capture) body.append("x")
            body.append(move.to.notation)
        }
        return body.toString() + suffix
    }

    private fun disambiguation(
        piece: Piece,
        move: DuckChessRules.Move,
        legalMoves: List<DuckChessRules.Move>,
        position: Position,
    ): String {
        val others = legalMoves.mapNotNull { candidate ->
            if (candidate.from == move.from || candidate.to != move.to) return@mapNotNull null
            val other = position.piece(candidate.from) ?: return@mapNotNull null
            if (other.kind != piece.kind || other.color != piece.color) return@mapNotNull null
            candidate.from
        }
        if (others.isEmpty()) return ""
        if (others.none { it.file == move.from.file }) return move.from.file.letter
        if (others.none { it.rank == move.from.rank }) return move.from.rank.value.toString()
        return move.from.notation
    }
}
